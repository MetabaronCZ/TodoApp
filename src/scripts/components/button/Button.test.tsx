import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';

import { Button } from 'components/button/Button';

import { icos } from 'modules/icos';
import { withMockedProviders } from 'test/component';

describe('components/button/Button', () => {
  it('should render <button> Button by default', () => {
    const onClick = jest.fn();
    const tree = render(withMockedProviders(<Button onClick={onClick} />));
    expect(tree.container).toMatchSnapshot();

    expect(onClick).not.toBeCalled();
    expect(tree.queryAllByRole('link').length).toEqual(0);
    expect(tree.queryAllByRole('button').length).toEqual(1);

    fireEvent.click(tree.getByRole('button'));
    expect(onClick).toBeCalledTimes(1);
    expect(onClick.mock.calls[0][0]).toEqual(undefined);
  });

  it('should render link Button when href defined', () => {
    const tree = render(withMockedProviders(<Button href="/test" />));
    expect(tree.container).toMatchSnapshot();

    expect(tree.queryAllByRole('link').length).toEqual(1);
    expect(tree.queryAllByRole('button').length).toEqual(0);
    expect(tree.getByRole('link')).toHaveAttribute('href', '/test');
  });

  it('should prefer link Button when href and onClick defined', () => {
    const onClick = jest.fn();

    const tree = render(
      withMockedProviders(<Button href="/test" onClick={onClick} />),
    );
    expect(tree.container).toMatchSnapshot();

    expect(onClick).not.toBeCalled();
    expect(tree.queryAllByRole('link').length).toEqual(1);
    expect(tree.queryAllByRole('button').length).toEqual(0);
    expect(tree.getByRole('link')).toHaveAttribute('href', '/test');

    fireEvent.click(tree.getByRole('link'));
    expect(onClick).toBeCalledTimes(1);
    expect(onClick.mock.calls[0][0]).toEqual(undefined);
  });

  it('should render props', () => {
    const { container } = render(
      withMockedProviders(
        <Button
          id="test-id"
          className="test-class"
          title="Test button"
          text="Button text"
        />,
      ),
    );
    expect(container).toMatchSnapshot();
    expect(container.firstChild).toHaveClass('test-class');
    expect(container.firstChild).toHaveAttribute('id', 'test-id');
    expect(container.firstChild).toHaveAttribute('title', 'Test button');
    expect(container.firstChild?.textContent).toEqual('Button text');
  });

  it('should render ico button', () => {
    const { container } = render(withMockedProviders(<Button ico="close" />));
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('i').length).toEqual(1);
    expect(container.textContent).toEqual(icos.close);
  });

  it('should render ico before', () => {
    const { container } = render(
      withMockedProviders(<Button ico="close" text="Button" />),
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('i').length).toEqual(1);
    expect(container.textContent).toEqual(`${icos.close}Button`);
  });

  it('should render ico after', () => {
    const { container } = render(
      withMockedProviders(<Button icoAfter="close" text="Button" />),
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('i').length).toEqual(1);
    expect(container.textContent).toEqual(`Button${icos.close}`);
  });

  it('should render both icos', () => {
    const { container } = render(
      withMockedProviders(
        <Button icoBefore="arrowLeft" icoAfter="arrowRight" text="Button" />,
      ),
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('i').length).toEqual(2);
    expect(container.textContent).toEqual(
      `${icos.arrowLeft}Button${icos.arrowRight}`,
    );
  });

  it('should render disabled state when onClick defined', () => {
    const onClick = jest.fn();

    const tree = render(
      withMockedProviders(<Button onClick={onClick} disabled />),
    );
    expect(onClick).not.toBeCalled();

    expect(tree.container).toMatchSnapshot();
    expect(tree.container.firstChild).toBeDisabled();

    const button = tree.getByRole('button');
    expect(button).toBeInTheDocument();

    fireEvent.click(button);
    expect(onClick).not.toBeCalled();
  });

  it('should render disabled state when href defined', () => {
    const onClick = jest.fn();

    const tree = render(
      withMockedProviders(<Button href="/test" onClick={onClick} disabled />),
    );
    expect(onClick).not.toBeCalled();

    expect(tree.container).toMatchSnapshot();
    expect(tree.container.firstChild).toBeDisabled();

    const button = tree.getByRole('button');
    expect(button).toBeInTheDocument();

    fireEvent.click(button);
    expect(onClick).not.toBeCalled();
  });
});
