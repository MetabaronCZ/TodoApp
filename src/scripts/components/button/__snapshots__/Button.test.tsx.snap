// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/button/Button should prefer link Button when href and onClick defined 1`] = `
.c0 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c0:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c0:disabled {
  color: #c0c0c0;
}

<div>
  <a
    class="c0"
    href="/test"
  />
</div>
`;

exports[`components/button/Button should render <button> Button by default 1`] = `
.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

<div>
  <button
    class="c0 c1"
    type="button"
  />
</div>
`;

exports[`components/button/Button should render both icos 1`] = `
.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <button
    class="c0 c1"
    type="button"
  >
    <i
      class="c2"
    >
      ❮
    </i>
    <span
      class="c3"
    >
      Button
    </span>
    <i
      class="c2"
    >
      ❯
    </i>
  </button>
</div>
`;

exports[`components/button/Button should render disabled state when href defined 1`] = `
.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

<div>
  <button
    class="c0 c1"
    disabled=""
    type="button"
  />
</div>
`;

exports[`components/button/Button should render disabled state when onClick defined 1`] = `
.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

<div>
  <button
    class="c0 c1"
    disabled=""
    type="button"
  />
</div>
`;

exports[`components/button/Button should render ico after 1`] = `
.c3 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <button
    class="c0 c1"
    type="button"
  >
    <span
      class="c2"
    >
      Button
    </span>
    <i
      class="c3"
    >
      ✖
    </i>
  </button>
</div>
`;

exports[`components/button/Button should render ico before 1`] = `
.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <button
    class="c0 c1"
    type="button"
  >
    <i
      class="c2"
    >
      ✖
    </i>
    <span
      class="c3"
    >
      Button
    </span>
  </button>
</div>
`;

exports[`components/button/Button should render ico button 1`] = `
.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

<div>
  <button
    class="c0 c1"
    type="button"
  >
    <i
      class="c2"
    >
      ✖
    </i>
  </button>
</div>
`;

exports[`components/button/Button should render link Button when href defined 1`] = `
.c0 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  padding-right: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c0:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c0:disabled {
  color: #c0c0c0;
}

<div>
  <a
    class="c0"
    href="/test"
  />
</div>
`;

exports[`components/button/Button should render props 1`] = `
.c0 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c1 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  background: #f0f0f0;
}

.c1:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c1:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c1:disabled {
  color: #c0c0c0;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <button
    class="c0 c1 test-class"
    id="test-id"
    title="Test button"
    type="button"
  >
    <span
      class="c2"
    >
      Button text
    </span>
  </button>
</div>
`;
